io.c

```diff
@@ -2244,6 +2237,186 @@ static int io_sqe_files_register(struct io_ring_ctx *ctx, void __user *arg,
 	return ret;
 }
 
+static int io_sqe_register_buffers(struct io_ring_ctx *ctx, void __user *uarg)
+{
+	struct vm_area_struct **vmas = NULL;
+	struct page **pages = NULL;
+	int buf_index, j;
+	int ret = -EINVAL;
+
+	struct io_mapped_ubuf *imu;
+	unsigned long off, start, end, ubuf;
+	int pret, nr_pages, max_nr_pages;
+	struct pxd_ioc_register_buffers arg;
+	size_t size;
+	bool existing = true;
+	long offset = 0;
+
+	ret = -EFAULT;
+	if (copy_from_user(&arg, uarg, sizeof(arg)))
+		goto err;
+
+	ret = -EINVAL;
+	if (arg.len & ~PAGE_MASK) {
+		pr_err("%s: mapped len is not aligned", __func__);
+		goto err;
+	}
+
+	if ((uintptr_t)arg.base & ~PAGE_MASK) {
+		pr_err("%s: mapped addr is not aligned", __func__);
+		goto err;
+	}
+
+	if (arg.max_len & ~PAGE_MASK) {
+		pr_err("%s: mapped max len is not aligned", __func__);
+		goto err;
+	}
+
+	if (arg.max_len < arg.len) {
+		pr_err("%s: max len less than len", __func__);
+		goto err;
+	}
+
+	for (buf_index = 0; buf_index < PXD_IO_MAX_USER_BUFS; ++buf_index) {
+		imu = &ctx->user_bufs[buf_index];
+		if (imu->ubuf <= (uintptr_t) arg.base &&
+		    (uintptr_t) arg.base + arg.len <= imu->ubuf + imu->len) {
+			offset = ((uintptr_t)arg.base - imu->ubuf) >> PAGE_SHIFT;
+			pr_info("found existing mapping %d, off %ld", buf_index, offset);
```

Comment:
if one registers a range <buf,len,max_len> that falls within an existing map, max_len arg is not checked to be within the same map; only the passed active buf range<addr,len> is checked and passes silently, but max_len also should fall within the same map, otherwise should not be accepted as a pass. 

I fail to understand, why support sparse range at all? At any point in time, caller goes through the same ioctl, with the newer args. It would be just plain efficient and simpler, if a new range needs to be  registered to call the ioctl to register the range. In any case, code has been extended to allow multiple maps to be registered at any point in time. max_len is superfluous and should be removed.

Maybe ranges can be coalesced internally, but that does not buy any advantage at all.

Reply:
Could you please show me the code which demonstrates how the ranges will be easily coalesced? If this is too much trouble for you let's proceed with the current implementation. You only comment is that interface is inconsistent. This does not seem like a show stopper to me. You can always make changes that will improve the interface later.